internal/repos/npm_packages.go

```diff
@@ -0,0 +1,115 @@
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/npmpackages"
+	"github.com/sourcegraph/sourcegraph/internal/jsonc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+// A NPMPackagesSource creates git repositories from `*-sources.tar.gz` files of
+// published NPM dependencies from the JS ecosystem.
+type NPMPackagesSource struct {
+	svc    *types.ExternalService
+	config *schema.NPMPackagesConnection
+}
+
+// NewNPMPackagesSource returns a new NPMSource from the given external
+// service.
+func NewNPMPackagesSource(svc *types.ExternalService) (*NPMPackagesSource, error) {
+	var c schema.NPMPackagesConnection
+	if err := jsonc.Unmarshal(svc.Config, &c); err != nil {
+		return nil, fmt.Errorf("external service id=%d config error: %s", svc.ID, err)
+	}
+	return &NPMPackagesSource{svc: svc, config: &c}, nil
+}
+
+var _ Source = &NPMPackagesSource{}
+
+// ListRepos returns all NPM artifacts accessible to all connections
+// configured in Sourcegraph via the external services configuration.
+func (s *NPMPackagesSource) ListRepos(ctx context.Context, results chan SourceResult) {
+	npmPackages, err := npmPackages(*s.config)
+	if err != nil {
+		results <- SourceResult{Err: err}
+		return
+	}
+	for _, npmPackage := range npmPackages {
+		repo := s.makeRepo(npmPackage)
+		results <- SourceResult{
+			Source: s,
+			Repo:   repo,
+		}
+	}
+	if err != nil {
+		results <- SourceResult{Err: err}
+		return
+	}
+	// TODO: [npm-package-support-database] Implement database code path here.
+}
+
+func (s *NPMPackagesSource) makeRepo(npmPackage reposource.NPMPackage) *types.Repo {
+	urn := s.svc.URN()
+	cloneURL := npmPackage.CloneURL()
+	repoName := npmPackage.RepoName()
+	return &types.Repo{
+		Name: repoName,
+		URI:  string(repoName),
+		ExternalRepo: api.ExternalRepoSpec{
+			ID:          string(repoName),
+			ServiceID:   extsvc.TypeNPMPackages,
+			ServiceType: extsvc.TypeNPMPackages,
+		},
+		Private: false,
+		Sources: map[string]*types.SourceInfo{
+			urn: {
+				ID:       urn,
+				CloneURL: cloneURL,
+			},
+		},
+		Metadata: &npmpackages.Metadata{
+			Package: npmPackage,
+		},
+	}
+}
+
+// ExternalServices returns a singleton slice containing the external service.
+func (s *NPMPackagesSource) ExternalServices() types.ExternalServices {
+	return types.ExternalServices{s.svc}
+}
+
+// npmPackages gets the list of applicable packages by de-duplicating dependencies
+// present in the configuration.
+func npmPackages(connection schema.NPMPackagesConnection) ([]reposource.NPMPackage, error) {
+	dependencies, err := npmDependencies(connection)
+	if err != nil {
+		return nil, err
+	}
+	npmPackages := []reposource.NPMPackage{}
+	isAdded := make(map[reposource.NPMPackage]bool)
+	for _, dep := range dependencies {
+		npmPackage := dep.NPMPackage
+		if _, added := isAdded[npmPackage]; !added {
```

Comment:
As a Go tip, this can be simplified to:
```suggestion
		if !isAdded[npmPackage] {
```

Reply:
Interesting, I didn't know that worked. I saw the [first answer on StackOverflow](https://stackoverflow.com/a/2050629/2682729), which suggests the first approach, and it is also marked as "Answer recommended by Go Language", but I now see that the second answer has the shorter code.